File: golang/2021-03/searchA2DMatrix.go
package _021_03

// Search a 2D Matrix
// leetCode: https://leetcode-cn.com/problems/search-a-2d-matrix/
// Write an efficient algorithm that searches for a value in an m x n matrix.
// This matrix has the following properties:
// Integers in each row are sorted from left to right.
// The first integer of each row is greater than the last integer of the previous row.

func SearchA2DMatrix(matrix [][]int, target int) bool {
	left := 0
	colNum := len(matrix)
	rowNum := len(matrix[0])
	right :=  colNum * rowNum - 1
	for left <= right {
		mid := (left + right) / 2
		col := mid / rowNum
		row := mid % rowNum
		midValue := matrix[col][row]
		if target == midValue {
			return true
		}else if target < midValue {
			right = mid - 1
		}else {
			left = mid + 1
		}
	}
	return false
}

File: rust/rust-2022/src/bin/searchA2DMatrix.rs
// Search a 2D Matrix
// leetcode: https://leetcode-cn.com/problems/search-a-2d-matrix-ii/
struct Solution {}

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let mut row = matrix.len() - 1;
        let mut r: i32 = row as i32;
        let mut col = 0;
        while r >= 0 && col < matrix[0].len() {
            if matrix[row][col] > target {
                r -= 1;
                if row > 0 {
                    row -= 1;
                }
            }else if matrix[row][col] < target {
                col += 1;
            }else {
                return true;
            }
        }
        false
    }
}

fn main() {
    let mut v = Vec::new();
    v.push(vec![-5]);
    let ans = Solution::search_matrix(v, -10);
    println!("{}", ans);
}